### ISSUE TITLE
Adjusting the intercept after downsampling

### ISSUE THREAD

**Issue by andrewpbray** [topepo/caret#994]
I've had good results with easing the estimation process for binary classification by using the downsampling routine outlined [in the book](https://topepo.github.io/caret/subsampling-for-class-imbalances.html#subsampling-during-resampling). The issue that I've come up against is that while using the `predict()` method for a test data set on the resulting `train` objects, I get a much higher general predicted base rate than if I hadn't downsampled.

My suspicion is that, in essence, the intercept term needs to be adjusted in account for the downsampling, but there may well be something else that I'm missing.

### Reproducible example:

The following is unfortunately not exactly a minimal example (it takes several minutes to run on my machine), but it was selected because it extends the example provided in the book. We compare the predicted expected value of the total number of cases over the test data set for the downsampled model (`down_inside`) and a non-downsampled model (`vanilla` - this is the one that takes awhile to run).

FWIW, my actual use-case is getting good predictions out of a lasso logistic model fit with `glmnet`.

``` r
library(caret)

set.seed(2969)
imbal_train <- twoClassSim(10000, intercept = -20, linearVars = 20)
imbal_test  <- twoClassSim(10000, intercept = -20, linearVars = 20)

ctrl_down <- trainControl(method = "repeatedcv", repeats = 5,
                     classProbs = TRUE,
                     summaryFunction = twoClassSummary,
                     sampling = "down")

set.seed(5627)
down_inside <- train(Class ~ ., data = imbal_train,
                     method = "treebag",
                     nbagg = 50,
                     metric = "ROC",
                     trControl = ctrl_down)

ctrl <- trainControl(method = "repeatedcv", repeats = 5,
                     classProbs = TRUE,
                     summaryFunction = twoClassSummary)

set.seed(5627)
vanilla <- train(Class ~ ., data = imbal_train,
                     method = "treebag",
                     nbagg = 50,
                     metric = "ROC",
                     trControl = ctrl)

p_down    <- predict(down_inside, imbal_test, type = "prob")$Class2
p_vanilla <- predict(vanilla, imbal_test, type = "prob")$Class2
sum(p_down)
#> [1] 2378.74
sum(p_vanilla)
#> [1] 567.28
```

Created on 2019-01-29 by the [reprex
package](http://reprex.tidyverse.org) (v0.2.0).

### Session Info:
```{r}
> sessionInfo()
R version 3.5.1 (2018-07-02)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Sierra 10.12.6

Matrix products: default
BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] reprex_0.2.0

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.0         rstudioapi_0.9.0   knitr_1.20         whisker_0.3-2      magrittr_1.5      
 [6] usethis_1.4.0.9000 devtools_2.0.1     pkgload_1.0.2      R6_2.3.0           rlang_0.3.1       
[11] stringr_1.3.1      tools_3.5.1        pkgbuild_1.0.2     sessioninfo_1.1.1  cli_1.0.1         
[16] clipr_0.5.0        withr_2.1.2        htmltools_0.3.6    remotes_2.0.2      yaml_2.2.0        
[21] assertthat_0.2.0   digest_0.6.18      rprojroot_1.3-2    crayon_1.3.4       processx_3.1.0    
[26] callr_2.0.4        fs_1.2.6           testthat_2.0.0     evaluate_0.11      memoise_1.1.0     
[31] glue_1.3.0         rmarkdown_1.10     stringi_1.2.4      compiler_3.5.1     desc_1.2.0        
[36] backports_1.1.3    prettyunits_1.0.2 
```

**Comment by topepo** [comment]
> My suspicion is that, in essence, the intercept term needs to be adjusted in account for the downsampling, but there may well be something else that I'm missing.

For logistic regression, down-sampling mainly adjusts the intercept. I'm told that there is an analytical proof of this somewhere but I have not seen it. In effect, this is what down-sampling is doing. 

The main effect that these sampling methods have on models is to calibrate (or un-calibrate if you are Bayesian) the class probabilities. For your example:
![image](https://user-images.githubusercontent.com/5731043/51927907-b2172680-23c2-11e9-961e-b90b8920ac66.png)

It has a habit of shrinking the minority class prediction more towards the extreme and, as a consequence, the 50% cutoff gives a more reasonable tradeoff between sensitivity and specificity.

**Comment by andrewpbray** [comment]
Yeah, that makes sense. I guess, then, my question is how to recalibrate (reuncalibrate?) those probabilities. In setting I'm working in, it's these probabilities that I need, not a thresholded binary decision.

The best I've been able to find is [this stackoverflow exchange](https://stackoverflow.com/questions/45164528/convert-predicted-probabilities-after-downsampling-to-actual-probabilities-in-cl) and the accompanying paper, but I figured I'd ping an issue to see if this was already implemented and I was just missing it.
